### ISSUE TITLE
intermittent spawnProcess failure in twisted.internet.test.test_process on Linux

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3571]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3571|
|Type|defect|
|Created|2008-12-10 15:56:06Z|
|Branch|https://github.com/twisted/twisted/tree/impatient-tests-3571|

Sometimes some of the tests in `twisted.internet.test.test_process` cause some bytes to be written to stderr from the child process created by spawnProcess from the catch-all exception handling path:

```
twisted.internet.test.test_process
  PTYProcessTestsBuilder_SelectReactor
    test_callWhenRunningSpawnProcessWarningFree ...                        [OK]
    test_processExitedWithSignal ...                                       [OK]
    test_spawnProcessTooEarlyWarns ... Upon execvpe /usr/bin/python ['/usr/bin/python', '-c', ''] in environment id 142718604
:Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/process.py", line 395, in _fork
    self._setupChild(**kwargs)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/process.py", line 859, in _setupChild
    fcntl.ioctl(slavefd, termios.TIOCSCTTY, '')
IOError: [Errno 5] Input/output error
```

This doesn't cause the tests to fail though, since there's no real error reporting to the parent process of this sort of failure.

This is probably due to a test bug where the tests don't wait for the child process to be fully initialized before letting the reactor cleanup code from `tearDown` run.  When the parent process closes the PTY, if the child hasn't yet done the `TIOCSCTTY` ioctl, it fails when it does happen.

The tests should wait for the process to exit before letting the cleanup code run.

<details><summary>Searchable metadata</summary>

```
trac-id__3571 3571
type__defect defect
reporter__exarkun exarkun
priority__low low
milestone__ 
branch__branches_impatient_tests_3571 branches/impatient-tests-3571
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1228924566000000 1228924566000000
changetime__1229526244000000 1229526244000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [25613]) Branching to 'impatient-tests-3571'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

This problem with the tests is fixed now.  I think there's still some issues, but they're hard to reproduce and probably not related to this other than being in the same code.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

By the way, the error output in the description is more easily and reliably produced by inserting a brief sleep (0.001s, say) into the very beginning of `PTYProcess._setupChild`.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

Great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [25700]) Merge impatient-tests-3571 

Author: exarkun
Reviewer: therve
Fixes: [#3571](https://github.com/twisted/twisted/issues/3571)

Change a couple `IReactorProcess` tests to make them more reliable.
`test_spawnProcessTooEarlyWarns` and `test_callWhenRunningSpawnProcessWarningFree`
now wait for the processes they spawn to complete before returning.
This prevents the PTY from being closed by the harness before the child process is
done with it.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
